### PR TITLE
WIP/Feature #4917 - active direcory realmprovider

### PR DESCRIPTION
**Pull request by helge000**
Adds a proposal for an active directory ream provider using [wirm](https://github.com/WinRb/WinRM)

As discussed on ML with @stbenjam, I mainly refactored FreeIPA as as active directory.
Currently I do not see a way on how to implement the OTP.
Also, it only provides a singe auth method (NTLM) and no https; more methodes will be implemented once we get further.
Same is true for test cases - but I am at it, @witlessbird and might need your help (interestingly it seems FreeIPA does not get tested at all).

Note, because this introduces `winrm`, we can run this provider on any host, just run:
`winrm quickconfig` as admin on a DC as admin.

### CODE REVIEW

File: bundler.d/windows.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 group :windows do
   gem 'highline', :platforms => [:mingw, :x64_mingw]
   gem 'win32-service', :platforms => [:mingw, :x64_mingw]
+  gem 'winrm', '~> 1.6'
```
Comment (helge000) [inline bundler.d/windows.rb:4]: Is this ok here?

**Comment by stbenjam** [comment]
>  Currently I do not see a way on how to implement the OTP.

AD supports pre-setting computer accounts.  If WinRM can't do it, then I think that's a big strike against using it, as not only does it require windows and powershell, it doesn't do the most useful thing for us.

adcli and mskutil support it: http://stef.thewalter.net/how-to-join-active-directory-domains.html

**Comment by helge000** [comment]
@stbenjam, let me rephrase: I certainly can set an Account Password. This is normally randomized,  240 char when password negotiated when the computer joints the domain.
By default, it is the same as the [SamAccountName](http://powershell.org/wp/forums/topic/reset-ad-computer-account/#post-11787).
I did not find anything on the topic on how to join a windows computer with this OTP or how this works with adcli (I cannot imagine adcli can do more than the AD powershell cmdlets?)

**Comment by stbenjam** [comment]
OTP means one-time password, and the link I just sent you clearly shows pre-adding a computer to AD, and then using that password to register a linux host using realmd.

What would the smart proxy provider be used for if you're not returning a password to Foreman?

**Comment by helge000** [comment]
When I researched in the behavior of adcli I read [this](http://stef.thewalter.net/how-to-join-active-directory-domains.html) on the matter. 
It basically states adlci is using a unsupported(?) behavior of active directory computer accounts; if so, it sets an account password and flags the account as 'change password on next login' or else. This certainly can be mimiced. Whatever we can use the same method to join windows computers needs more testing.

**Comment by helge000** [comment]
Can you show me the part where FreeIPA returns the OTP, @stbenjam? I would add the function.

**Comment by stbenjam** [comment]
It's included in the JSON return of the API request: https://github.com/theforeman/smart-proxy/blob/develop/modules/realm/freeipa.rb#L117

It's in a 'randompassword' hash key, and used here: https://github.com/theforeman/foreman/blob/develop/app/models/concerns/orchestration/realm.rb#L35

It's still not clear to me why you would pre-add the host to active driectory if not to do this kind of thing.  What benefit does an AD provider get you if it's not providing a mechanism for hosts to securely register?

IHMO that's the entire reason we have this feature in Foreman.

**Comment by helge000** [comment]
> IHMO that's the entire reason we have this feature in Foreman.
> @stbenjam Right; though at least for windows hosts there are some more like GPO's

I implemented the mentioned OTP together with a better method for rebuild; by just resetting the account password to our OTP. 

This would be a sample return from rebuild, hope this works with [Foreman:](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/orchestration/realm.rb#L35)

``` json
{
  "UserPrincipalName": null,
  "Enabled": true,
  "SamAccountName": "test03$",
  "SID": {
    "BinaryLength": 28,
    "AccountDomainSid": {
      "BinaryLength": 24,
      "AccountDomainSid": "S-1-5-21-335421253-545455696-1376732018",
      "Value": "S-1-5-21-335421253-545455696-1376732018"
    },
    "Value": "S-1-5-21-335421253-545455696-1376732018-1292"
  },
  "DistinguishedName": "CN=test03,CN=Computers,DC=int,DC=domain,DC=com",
  "Name": "test03",
  "ObjectClass": "computer",
  "ObjectGuid": "45bbbdd6-9a80-40eb-a352-016afa4e7655",
  "PropertyNames": [
    "DistinguishedName",
    "Enabled",
    "Name",
    "ObjectClass",
    "ObjectGUID",
    "SamAccountName",
    "SID",
    "UserPrincipalName"
  ],
  "AddedProperties": [

  ],
  "RemovedProperties": [

  ],
  "ModifiedProperties": [

  ],
  "PropertyCount": 8,
  "ObjectGUID": "45bbbdd6-9a80-40eb-a352-016afa4e7655",
  "password": "mxckxphbospolcdwscmnfdnjcbijtiehkspkmnppsdmnrdacalwxfullclaouqrgzwwdpbtkktveemxhxdgytlyyllhgmmtkghw"
}
```

**Comment by stbenjam** [comment]
Should be `randompassword` in the hash, but otherwise yea.  Do you have some instructions for how to test this? I don't think we have much AD knowledge among Foreman developers.

I can probably find some time to spin up a directory controller eventually, but not sure exactly what needs to be done to get this and WinRM working.

**Comment by helge000** [comment]
> Should be randompassword

Right... already fixed that locally.

Can you explain the foreman behavior for me [here](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/orchestration/realm.rb#L28) and how such a options hash from foreman would look like? For instance, do I get `options[:update] = true`? The example I found [in the wiki](http://projects.theforeman.org/projects/smart-proxy/wiki/API) does not mention such a thing; it would certainly be useful. I figure for example:

``` ruby
{ options => {
  :rebuild => false, 
  :hostname => 'foo.bar.com', 
  :userclass => 'myhostgroup', 
  :update => true
  }
}
```

Also, I assume `hostgroup.title` is the name of the hostgroup and not the full path including nested?

> can probably find some time to spin up a directory controller eventually, but not sure exactly what needs to be done to get this and WinRM working.

You might want to have a look at the [integration tests WinRM is using](https://github.com/WinRb/WinRM#running-the-tests); basically a variant box. Also, your GitHub profile says you're Berlin based, @stbenjam; if you like I can help out using our infra.

**Comment by dmitri-d** [comment]
I think it might be helpful to introduce providers for the realm module. Would any of you be up for that? Alternatively, I can have something by the end of the week (feeling sick today)...

**Comment by helge000** [comment]
> I think it might be helpful to introduce providers

@witlessbird: defiantly. Though I'm pretty sure this would beyond me. I am happy to rebase :). 
BTW, where do I best put the gem in bunlder.d? Currently it is in windows.rb.

Also, I am unable to figure why Jenkins throws rubocop errors; I am running it locally and there are no offences. Off topic, you merged #369 and one test unit suddenly also [fails](http://ci.theforeman.org/job/test_proxy_develop_pr_rubocop/587/checkstyleResult/package.-371464988/source.3/#5). Should I fix this with a new PR, because it will fail every build now I suppose.

Get well soon!

**Comment by stbenjam** [comment]
> For instance, do I get options[:update] = true? The example I found in the wiki does not mention such a thing; it would certainly be useful. I figure for example:

Options is just a variable name in Foreman.  Those get sent as parameters, which are available in `params`, see https://github.com/theforeman/smart-proxy/blob/develop/modules/realm/freeipa.rb#L95
for how it access things

> You might want to have a look at the integration tests WinRM is using; basically a variant box. Also, your GitHub profile says you're Berlin based, @stbenjam; if you like I can help out using our infra.

Thanks I'll have a look.  Location-wise, that needs updating, I just moved a few months ago unfortunately (to the US).

**Comment by helge000** [comment]
> Location-wise, that needs updating, I just moved a few months ago unfortunately (to the US).

A pitty ;)

> Options is just a variable name in Foreman. Those get sent as parameters,

That's what I meant.. So I have basically the params hash with `:hostname`, `:userclass`, `:rebuild`, `:update`? 
On any rebuild and update and except delete Foreman calls POST?

(I am asking because I can only test test the smart-proxy directly using curl and did not have a chance to get this running with Foreman.)

**Comment by stbenjam** [comment]
> That's what I meant.. So I have basically the params hash with :hostname, :userclass, :rebuild, :update? On any rebuild and update and except delete Foreman calls POST?

Yup, exactly

File: bundler.d/windows.rb (hunk 1 above)
Comment (dmitri-d) [inline bundler.d/windows.rb:4]: I think so.

**Comment by dmitri-d** [comment]
test failures are due to new dependencies being ruby >= 1.9.3 only. The limitation is ok -- smart-proxy on Windows is also ruby 1.9.3 and higher only, but to get the build passing may need to put those new dependencies into a dedicated bundler group, which would be easier if this functionality used providers...

**Comment by helge000** [comment]
> put those new dependencies into a dedicated bundler group, which would be easier if this functionality used providers.

@witlessbird - a little too late I had a change to watch [your deep dive](https://www.youtube.com/watch?v=fEMB1EqzQG4) - and was deciding to exactly that. The code is a little mess ATM as I accidentally pushed some not yet finished stuff while working on #379 

Do you think I should rather put it in a plugin?

I'll continue with this once the #379  thing is done. Also, you mentioned something about braking realm down to providers? As the 1.11 branching is done, this might be a good time :)

**Comment by dmitri-d** [comment]
Changing realm module to use providers is definitely a way to go. I may have some time later this week, we shall see.

**Comment by dmitri-d** [comment]
@helge000: Are you planning to continue working on this PR?

**Comment by dmitri-d** [comment]
@helge000: I'd like to close this PR as there's now an AD realm provider available, unless you want to continue with your approach?
